### ISSUE TITLE
Share C# code between .cshtml Files

### ISSUE THREAD

**Issue by DNF-SaS** [statiqdev/Statiq#631]
Hi,

I'd like to write a small translation-Helper (basically a "pure" function, which simply tries translate a string provided as parameter). Of course, it should be available for all cshtml-files.
However, I cannot make it work - can anybody point out, how to do this in wyam razor?
(App_Code does not work AFAIK)

Thanks,
Sascha

**Comment by avishnyakov** [comment]
Hey @DNF-SaS , we've got the same problem, described here in more details:
> Make a way to share, reuse wyam configuration files #626

A few workarounds we came up with are:
* move all helpers into Wyam config
* move all helpers into custom C# assembly and the load it in Wyam config (so that it would be exposed to Razor pages)

Hope that helps a bit. Also, keen on a better way to handle this. Especially with lack of reusability with built-in Razor - lack of app_code support and helpers, only functions-per-page.

**Comment by daveaglick** [comment]
As @avishnyakov said, the recommended way to expose helpers to Razor in Wyam is to add them to your config file. Any methods in the config file will be available within the Razor file. Generally you’ll want to add them to a static helper class in the config file - you can even make them extension methods of `HtmlHelper`.

Right now we’re stuck with the version and capabilities of the Razor implemention currently in Wyam. They changed a lot of the API with the next version of Razor and Wyam won’t be able to update to it until the .NET Core port is done later this year.

**Comment by avishnyakov** [comment]
@daveaglick is it possible to have additional preprocessor directive or Wyam API to load C# helpers/classes from `.cs` files outside of Wyam config? 

That would be wonderful. Nice split up between Wyam config and helpers, Wyam config and helpers re-usability due to de-coupling, and then imagine that themes would be able to ship these csharp helpers - so that wyam config would never change, all changes would be at theme/cs files level.

Another idea - maybe even not a directive, but a new module which would read target *.cs files, evaluate them and load into the current assembly domain. I think we can do this right now with Execute() module, evaluating stuff on the fly. That would work well with Wyam restarts, not sure how to figure out live reload due to uplifting new assemblies into the same domain and id they would be promoted to Razor by default. Just thinking aloud.

**Comment by daveaglick** [comment]
Another directive just for .cs files is certainly possible, but I’d rather wait a little longer and do this the right way by loading other config files since they’re a superset of C# - once that gets in, it’ll work for both .cs files and config files.

The `Execute` module isn’t quite the same thing - it invokes a delegate that you’ve defined and has been compiled in the config file. You’re probably thinking of a module that would take a string containing C# code and run it through Roslyn to compile/link it at runtime - no such module exists right now and I’m not sure what utility there would be in that over a pre-compiled delegate like `Execute` uses.

You also mentioned themes and I think that’s where all this comes together. Once we can load additional config files and get the ordering and everything correct, then themes can include their own config files that enhance the recipe for that specific theme. Could be a lot of neat possibilities with that.

**Comment by DNF-SaS** [comment]
Right now, I'd put the code into wyam.config.
But: how can I access the settings-object?
Wyam keeps saying "CS0122: 'Settings' is inaccessible due to its protection level".
Can anybody help?

**Comment by jonasdoerr** [comment]
Try `Context.Settings`. See [`IExecutionContext`](https://wyam.io/api/wyam.common.execution/iexecutioncontext/) for all members.

**Comment by DNF-SaS** [comment]
Thanks for the tip. Unfortunately, a static class defined in wyam.config has no access to settings.
Accessing settings in a Razor-View and passing it to the static class' function works. So it's okay.

**Comment by daveaglick** [comment]
@DNF-SaS Correct, any static classes in the config file won't have access to settings. That's because the code in the config file is actually in an instance method and static classes get "lifted" out of it. I.e.,

```
// Some config code here

public static class MyHelper
{
    public string SomeHelper()
    {
        // Some helper code
    }
}
```

Becomes:


```
public class ScriptBase
{
    public ISettings Settings { get; }
    // Other config script globals

    public void Run()
    {
        // Some config code here
    }
}

public static class MyHelper
{
    public string SomeHelper(string str)
    {
        // Some helper code
    }
}
```

The best way around this if you need access to the settings in the static helper class is to pass in the `IReadOnlySettings` instance when you call the helper method, which is available from the `IExecutionContext` available in your Razor or module code. I.e.:

```
public static class MyHelper
{
    public string SomeHelper(string str, IReadOnlySettings settings)
    {
        // Some helper code
    }
}
```
